### PR TITLE
Onboarding flow refactor (EXPOSUREAPP-2995)

### PR DESCRIPTION
**Pull request by SamuraiKek**
## Description

This PR introduces the MVVM pattern with dependency injection for the onboarding flow. It also introduces the code skeleton for running instrumentation tests for each of the fragments in the onboarding flow.

## How to test

 Run the app and click around the onboarding flow. No behaviour changes should be noticeable. Make sure you delete the app's data before running if you've already been using the app, otherwise the onboarding flow won't show up.

### CODE REVIEW

**Comment by harambasicluka** [comment]
Please remove the checklist & add a milestone :)

**Comment by SamuraiKek** [comment]
> Please remove the checklist & add a milestone :)

Yep, I will once I remove this PR from draft.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragment.kt [hunk 1]
```diff
@@ -10,34 +10,53 @@ import androidx.navigation.fragment.findNavController
...
+        vm.routeToScreen.observe2(this) {
+            when (it) {
+                is OnboardingNavigationEvents.NavigateToOnboardingPrivacy ->
+                    navigateToOnboardingPrivacyFragment()
+                is OnboardingNavigationEvents.NavigateToEasyLanguageUrl ->
+                    navigateToEasyLanguageUrl()
+            }
         }
-        setLinks()
+        setButtonOnClickListener()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragment.kt:40]: Inline this, maybe use
```
binding.apply {

}
```
The extra method reduces code readability.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragment.kt [hunk 2]
```diff
@@ -10,34 +10,53 @@ import androidx.navigation.fragment.findNavController
...
+        binding.onboardingInclude.onboardingEasyLanguage.setOnClickListener { vm.onEasyLanguageClick() }
+    }
+
+    private fun navigateToOnboardingPrivacyFragment() {
+        findNavController().doNavigate(
+            OnboardingFragmentDirections.actionOnboardingFragmentToOnboardingPrivacyFragment()
+        )
+    }
+
+    private fun navigateToEasyLanguageUrl() {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragment.kt:54]: What do you think about also inlining this again? It's short and sweet, you can maybe use the extension `Fragment.doNavigate` to make it even shorter. :smile:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragmentViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import com.squareup.inject.assisted.AssistedInject
+import de.rki.coronawarnapp.ui.SingleLiveEvent
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.SimpleCWAViewModelFactory
+
+class OnboardingFragmentViewModel @AssistedInject constructor() : CWAViewModel() {
+    val routeToScreen: SingleLiveEvent<OnboardingNavigationEvents> = SingleLiveEvent()
+
+    fun onNextButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingPrivacy
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragmentViewModel.kt:12]: Please use `postValue(...)`, which is async and means there are less requirements on the caller (which thread it is being called from).

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragmentViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    fun onEasyLanguageClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToEasyLanguageUrl
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingFragmentViewModel.kt:16]: Please use `postValue(...)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsFragment.kt [hunk 5]
```diff
@@ -31,15 +47,15 @@ class OnboardingNotificationsFragment : Fragment(R.layout.fragment_onboarding_no
     }
 
     private fun setButtonOnClickListener() {
-        binding.onboardingButtonNext.setOnClickListener {
-            navigateToMain()
-        }
-        binding.onboardingButtonBack.buttonIcon.setOnClickListener {
-            (activity as OnboardingActivity).goBack()
-        }
+        binding.onboardingButtonNext.setOnClickListener { vm.onNextButtonClick() }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsFragment.kt:50]: Inline this please.
If you want to extract something, extract it "by feature", i.e. do 
`setSetupOnboarding` which does buttons + observation.

But don't do `setAllButtons`, we want "package by feature".

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsFragment.kt [hunk 6]
```diff
@@ -31,15 +47,15 @@ class OnboardingNotificationsFragment : Fragment(R.layout.fragment_onboarding_no
...
+        binding.onboardingButtonBack.buttonIcon.setOnClickListener { vm.onBackButtonClick() }
     }
 
     private fun navigateToMain() {
         (requireActivity() as OnboardingActivity).completeOnboarding()
     }
+
+    private fun navigateToOnboardingTestFragment() {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsFragment.kt:58]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+import com.squareup.inject.assisted.AssistedInject
+import de.rki.coronawarnapp.ui.SingleLiveEvent
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.SimpleCWAViewModelFactory
+
+class OnboardingNotificationsViewModel @AssistedInject constructor() : CWAViewModel() {
+    val routeToScreen: SingleLiveEvent<OnboardingNavigationEvents> = SingleLiveEvent()
+
+    fun onNextButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToMainActivity
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsViewModel.kt:12]: `postValue(...)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    fun onBackButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingTracing
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsViewModel.kt:16]: `postValue(...)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragment.kt [hunk 9]
```diff
@@ -28,13 +44,17 @@ class OnboardingPrivacyFragment : Fragment(R.layout.fragment_onboarding_privacy)
...
     private fun setButtonOnClickListener() {
-        binding.onboardingButtonNext.setOnClickListener {
-            findNavController().doNavigate(
-                OnboardingPrivacyFragmentDirections.actionOnboardingPrivacyFragmentToOnboardingTracingFragment()
-            )
-        }
-        binding.onboardingButtonBack.buttonIcon.setOnClickListener {
-            (activity as OnboardingActivity).goBack()
-        }
+        binding.onboardingButtonNext.setOnClickListener { vm.onNextButtonClick() }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragment.kt:47]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragment.kt [hunk 10]
```diff
@@ -28,13 +44,17 @@ class OnboardingPrivacyFragment : Fragment(R.layout.fragment_onboarding_privacy)
...
+        binding.onboardingButtonBack.buttonIcon.setOnClickListener { vm.onBackButtonClick() }
+    }
+
+    private fun navigateToOnboardingTracing() {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragment.kt:51]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragment.kt [hunk 11]
```diff
@@ -28,13 +44,17 @@ class OnboardingPrivacyFragment : Fragment(R.layout.fragment_onboarding_privacy)
...
+        findNavController().doNavigate(
+            OnboardingPrivacyFragmentDirections.actionOnboardingPrivacyFragmentToOnboardingTracingFragment()
+        )
+    }
+
+    private fun navigateToOnboardingFragment() {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragment.kt:57]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+import com.squareup.inject.assisted.AssistedInject
+import de.rki.coronawarnapp.ui.SingleLiveEvent
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.SimpleCWAViewModelFactory
+
+class OnboardingPrivacyViewModel @AssistedInject constructor() : CWAViewModel() {
+    val routeToScreen: SingleLiveEvent<OnboardingNavigationEvents> = SingleLiveEvent()
+
+    fun onNextButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingTracing
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyViewModel.kt:12]: `postValue(...)` please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    fun onBackButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingFragment
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyViewModel.kt:16]: `postValue(...)` please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragment.kt [hunk 14]
```diff
@@ -27,13 +44,17 @@ class OnboardingTestFragment : Fragment(R.layout.fragment_onboarding_test) {
...
     private fun setButtonOnClickListener() {
-        binding.onboardingButtonNext.setOnClickListener {
-            findNavController().doNavigate(
-                OnboardingTestFragmentDirections.actionOnboardingTestFragmentToOnboardingNotificationsFragment()
-            )
-        }
-        binding.onboardingButtonBack.buttonIcon.setOnClickListener {
-            (activity as OnboardingActivity).goBack()
-        }
+        binding.onboardingButtonNext.setOnClickListener { vm.onNextButtonClick() }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragment.kt:47]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragment.kt [hunk 15]
```diff
@@ -27,13 +44,17 @@ class OnboardingTestFragment : Fragment(R.layout.fragment_onboarding_test) {
...
+        binding.onboardingButtonBack.buttonIcon.setOnClickListener { vm.onBackButtonClick() }
+    }
+
+    private fun navigateToOnboardingNotificationsFragment() {
+        findNavController().doNavigate(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragment.kt:52]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragment.kt [hunk 16]
```diff
@@ -27,13 +44,17 @@ class OnboardingTestFragment : Fragment(R.layout.fragment_onboarding_test) {
...
+            OnboardingTestFragmentDirections.actionOnboardingTestFragmentToOnboardingNotificationsFragment()
+        )
+    }
+
+    private fun navigateToOnboardingTracingFragment() {
+        (activity as OnboardingActivity).goBack()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragment.kt:58]: Inline please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestViewModel.kt [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+import com.squareup.inject.assisted.AssistedInject
+import de.rki.coronawarnapp.ui.SingleLiveEvent
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
+import de.rki.coronawarnapp.util.viewmodel.SimpleCWAViewModelFactory
+
+class OnboardingTestViewModel @AssistedInject constructor() : CWAViewModel() {
+    val routeToScreen: SingleLiveEvent<OnboardingNavigationEvents> = SingleLiveEvent()
+
+    fun onNextButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingNotifications
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestViewModel.kt:12]: `postValue(...)` please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestViewModel.kt [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    fun onBackButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingTracing
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestViewModel.kt:16]: `postValue(...)` please.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragment.kt [hunk 19]
```diff
@@ -63,19 +73,13 @@ class OnboardingTracingFragment : Fragment(R.layout.fragment_onboarding_tracing)
...
-        binding.onboardingButtonNext.setOnClickListener {
-            internalExposureNotificationPermissionHelper.requestPermissionToStartTracing()
-        }
-        binding.onboardingButtonDisable.setOnClickListener {
-            showCancelDialog()
-        }
-        binding.onboardingButtonBack.buttonIcon.setOnClickListener {
-            (requireActivity() as OnboardingActivity).goBack()
-        }
+        binding.onboardingButtonNext.setOnClickListener { vm.onNextButtonClick() }
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragment.kt:76]: Inline `setButtonOnClickListener` please, maybe use `binding.apply`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt [hunk 20]
```diff
@@ -40,6 +42,18 @@ class OnboardingTracingFragmentViewModel @AssistedInject constructor(
         }
     }
 
+    fun onNextButtonClick() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingTest
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt:46]: `postValue(...)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt [hunk 21]
```diff
@@ -40,6 +42,18 @@ class OnboardingTracingFragmentViewModel @AssistedInject constructor(
...
+    }
+
+    fun showCancelDialog() {
+        routeToScreen.value = OnboardingNavigationEvents.ShowCancelDialog
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt:50]: `postValue(...)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt [hunk 22]
```diff
@@ -40,6 +42,18 @@ class OnboardingTracingFragmentViewModel @AssistedInject constructor(
...
+    }
+
+    fun onBackButtonPress() {
+        routeToScreen.value = OnboardingNavigationEvents.NavigateToOnboardingPrivacy
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentViewModel.kt:54]: `postValue(...)`

File: Corona-Warn-App/src/main/res/layout/fragment_onboarding.xml [hunk 23]
```diff
@@ -2,6 +2,11 @@
 <layout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
     xmlns:tools="http://schemas.android.com/tools">
+    <data>
+        <variable
+            name="onboardingFragmentViewModel"
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/layout/fragment_onboarding.xml:7]: Why do we need the viewModel here?

File: Corona-Warn-App/src/main/res/layout/fragment_onboarding_notifications.xml [hunk 24]
```diff
@@ -1,6 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
 <layout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto">
+    <data>
+        <variable
+            name="onboardingNotificationsViewModel"
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/layout/fragment_onboarding_notifications.xml:6]: Why do we need the viewModel here?

File: Corona-Warn-App/src/main/res/layout/fragment_onboarding_privacy.xml [hunk 25]
```diff
@@ -8,6 +8,10 @@
 
         <import type="de.rki.coronawarnapp.R" />
 
+        <variable
+            name="onboardingPrivacyViewModel"
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/layout/fragment_onboarding_privacy.xml:12]: Why do we need the viewModel here?

File: Corona-Warn-App/src/main/res/layout/fragment_onboarding_test.xml [hunk 26]
```diff
@@ -1,6 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
 <layout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto">
+    <data>
+        <variable
```
Comment (d4rken) [inline Corona-Warn-App/src/main/res/layout/fragment_onboarding_test.xml:5]: Why do we need the viewModel here?

**Review by d4rken — CHANGES_REQUESTED** [review]
Mostly good. A bunch of small optimizations. I tried to explain but if you are still unclear about the why, please ask.

File: Corona-Warn-App/src/main/res/layout/fragment_onboarding.xml (hunk 23 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/main/res/layout/fragment_onboarding.xml:7]: My bad, I was under the wrong impression that we need to explicitly specify the viewModel inside the fragment layout. I'll remove them

**Review by d4rken — COMMENTED** [review]
lgtm

**Review by d4rken — DISMISSED** [review]

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentTest.kt [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+    fun launch_fragment() {
+        launchFragment<OnboardingNotificationsFragment>()
+    }
+}
+
+@Module
+abstract class OnboardingTracingFragmentTestModule {
+    @ContributesAndroidInjector
+    abstract fun onboardingTracingFragment(): OnboardingTracingFragment
+}
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentTest.kt:44]: minor nitpick: Some files are missing new lines

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragmentTest.kt [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+    fun launch_fragment() {
+        launchFragment<OnboardingTestFragment>()
+    }
+}
+
+@Module
+abstract class OnboardingTestFragmentModule {
+    @ContributesAndroidInjector
+    abstract fun onboardingTestFragment(): OnboardingTestFragment
+}
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTestFragmentTest.kt:44]: minor nitpick: Some files are missing new lines

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragmentTest.kt [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+    fun launch_fragment() {
+        launchFragment<OnboardingPrivacyFragment>()
+    }
+}
+
+@Module
+abstract class OnboardingPrivacyTestModule {
+    @ContributesAndroidInjector
+    abstract fun onboardingPrivacyFragment(): OnboardingPrivacyFragment
+}
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingPrivacyFragmentTest.kt:44]: minor nitpick: Some files are missing new lines

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsFragmentTest.kt [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+    fun launch_fragment() {
+        launchFragment<OnboardingNotificationsFragment>()
+    }
+}
+
+@Module
+abstract class OnboardingNotificationsTestModule {
+    @ContributesAndroidInjector
+    abstract fun onboardingNotificationsFragment(): OnboardingNotificationsFragment
+}
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingNotificationsFragmentTest.kt:44]: minor nitpick: Some files are missing new lines

**Review by kolyaopahle — DISMISSED** [review]
one very minor nitpick otherwise lgtm

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentTest.kt (hunk 27 above)
Comment (SamuraiKek) [reply Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/onboarding/OnboardingTracingFragmentTest.kt:44]: Ok, I'll add them

**Review by kolyaopahle — APPROVED** [review]

**Review by Oliver-Zimmerman — APPROVED** [review]
Code looks good to me and the onboarding flow was completed successfully when I tested it. I like how you handled navigation with the OnboardingNavigationEvents. I may adopt this in my Submission flow refactor.

**Pull request merged**